### PR TITLE
Always put OAuth user info in the ExternalLoginUser table

### LINKED ISSUE #1124: OAuth2 registered accounts should still be local by using ExternalLoginUser
I think this is important to fix before 1.1.0 final.
The current code handling OAuth2 users registration encodes the OAuth2 LoginSource in the user table, while still assigning a username/password pair to the newly create user that is supposed to be usable for "local" authentication. This is an inconsistent way to use that LoginSource field because then you would not know what username/password refer to unless by adding special code to handle "OAuth2" source. Indeed this is the special code added, for example in ForgotPasswordPost:
```
        if !u.IsLocal() && !u.IsOAuth2() {                                      
                ctx.Data["Err_Email"] = true                                    
                ctx.RenderWithErr(ctx.Tr("auth.non_local_account"), tplForgotPassword, nil)
                return                                                          
        }      
```

That's weird, because the OAuth2 user's password *is* indeed a "local" password,
so why should there be a difference ? The code also added an IsOauth2 method to the User table, building on what seems to be a broken model.

Users should not *BE* local or Oauth2, *credentials* should be. Users can be recognized by multiple credentials, but we'd always know about them locally, so they are all local. 

The inconsistency is confirmed by the fact that you can still associate an OAuth2 credential to your previosuly-registered account resulting in a "IsLocal" user with exactly the same caracteristic of the "IsOAuth2" user, except the "LoginSource" is different. Characteristics being: you can login with local username/password as well as with "OAuth2" credentials.

My suggestion here is to always keep the "LoginSource" field as an indication of the semantic for the "LoginSource" and "Password" fields in the "User" table, so to never use an OAuth2 LoginSource identifier in that table.

Can you see any drawback in that ?

\cc @willemvd

### PR DESCRIPTION
**Pull request by strk**
Never set OAuth in User.LoginSource, which keeps referring to
the username/password interpretation. Fixes #1124

\cc @willemvd, @morrildl, @lunny, @bkcsoft

I think this should be done before 1.1.0

### CODE REVIEW

**Comment by lunny** [comment]
It's really to need a proposal document to describe all the login.

**Comment by strk** [comment]
If a document is needed, it should be produced *before* introducing
the OAuth2 support, rather than let it add confusion to the code base
(and changes to production databases)

**Comment by strk** [comment]
Note that doing this *after* 1.1.0 is out will require a new migration to convert existing records with OAuth2 in User.LoginSource

**Comment by pgaskin** [comment]
@strk but even if it is done before, what happens to the installs from master builds (such as my own)?

**Comment by strk** [comment]
@geek1011 good question, I guess those would be just left with inconsistent database values. You might end up having users with OAuth2 LoginSource and no way to login with those OAuth2.

File: routers/user/auth.go
Comment (lunny) [inline routers/user/auth.go]: LoginType should be plain?

File: routers/user/auth.go
Comment (lunny) [inline routers/user/auth.go:916]: How to migrate the old user?

**Review by lunny — CHANGES_REQUESTED** [review]

**Comment by lunny** [comment]
I like this idea. There are two kinds of external Login Source. One is provide password authentication, another is for link.

File: routers/user/auth.go
Comment (strk) [reply routers/user/auth.go]: The default is LoginNone, which is threated by code the same as LoginPlain.
But yes, I guess it could be LoginPlain as long as we guarantee the actual login is not of the LDAP or SMTP or PAM type...

File: routers/user/auth.go
Comment (strk) [reply routers/user/auth.go:916]: Now that 1.1.0 is out yes, this is a problem. Too bad you didn't want to merge this before (we saw people had problems with early upgrades anyway, so this would have not been different)

Comment (lunny) [reply routers/user/auth.go:916]: Just like @geek1011 said, there many people use master, a migration needed.

**Comment by strk** [comment]
I raised this issue in the OAuth2 PR before, but my concern was not
considered. Now I don't know if I have time to fix this myself.
I think it should be done by the author of the OAuth2 code.

It's great to be more "open" in development, but we all need to pay
more attention to contributions and give proper importance to
responsibility. Whoever introduces a feature should be available to
maintain it.

**Comment by lunny** [comment]
@strk, when you review a PR, if you think your opinion should not be ignored, you could use github's request changes, then the merger should ask your confirmation before merge it. Our rules only consider two reviewers, lost this rule.

**Comment by lunny** [comment]
I will try to add migration for this PR when I have time.

**Comment by bkcsoft** [comment]
@lunny Did you have time to work on this? Otherwise I suggest we close this as "out of date"

**Comment by lunny** [comment]
@bkcsoft I'm afraid I have no time to work on this. I still have serval PRs in v1.3 and v1.4 need work.

**Comment by coolaj86** [comment]
It looks like this is related to https://github.com/go-gitea/gitea/issues/4226 and https://github.com/go-gitea/gitea/pull/5029

**Comment by techknowlogick** [comment]
Closing as outdated. Please re-open if needed.

**Comment by strk** [comment]
Note I CANNOT reopen. I can not modify any github ticket, although I'm supposedly a "maintainer" (I think this should change)

**Comment by strk** [comment]
Thanks @lunny, I'm force-pushing a version with conflicts resolved (but didn't try it again)

**Comment by strk** [comment]
@zeripath I've formatted the code, to make Drone CI happy (hopefully) -- in case you want to give this branch a try

**Comment by strk** [comment]
Still relevant, not sure why Travis is sad

**Comment by lafriks** [comment]
As per integration tests there is error in template:
```
template: user/settings/account:10:45: executing "user/settings/account" at <.SignedUser.IsOAuth2>: can't evaluate field IsOAuth2 in type interface {}
```

**Comment by singuliere** [comment]
Refactors moved  `routers/auth.go` to `routers/user/auth.go` and the `IsOAuth2` method from `models/user.go` to `models/user/user.go` but the logic changed by this PR remains the same. @strk it looks like your initial motivation for creating this PR is still valid, right?

**Comment by strk** [comment]
Yes, the motivation remains valid. Unfortunately the high version of Go required to build Gitea made me in a position to have more problems contributing (can't build w/out installing a foreign package in my otherwise all-distro-based system, and I'm short on disk space). For this reason I can't test my change much. Will gladly accept a PR to my branch if you want to work on the refactoring porting.

**Comment by wxiaoguang** [comment]
It's stale for petty long time.

What's the root problem behind the issue/PR?

Is there any valuable information in this context?

**Comment by strk** [comment]
It was mostly a design concern, a way to separate "external" authentication sources from "internal" ones. One problem being resolved would be described in https://github.com/go-gitea/gitea/issues/1124#issuecomment-284766818 but that issue was closed by stale bot too

**Comment by wxiaoguang** [comment]
Could some maintainers help to re-open that issue? @techknowlogick @lunny

**Comment by wxiaoguang** [comment]
I think this PR/issue could be kept open, until there is a clear conclusion

**Comment by lunny** [comment]
https://github.com/go-gitea/gitea/issues/1124 is opened so I think we could close this one because it's outdated.

**Comment by denyskon** [comment]
I also think we can close this and keep the issue open for further discussion

**Pull request closed without merging**
